### PR TITLE
feature(sdk): implement watch_status method

### PR DESCRIPTION
**Pull request by droctothorpe**
Co-authored-by: andreafehrman <andrea.k.fehrman@vanderbilt.edu>

**Description of your changes:**
This PR implements a `watch_status` method to the `Client` class that lets you monitor the progress of a run using a simple terminal UI provided by the [rich](https://github.com/Textualize/rich) library.

Here's a gif illustrating UX (please bear in mind that it's sped up 2.5x):
progress_upstream

This method also properly renders in a Jupyter notebook cell.

We have a lot of other capabilities (including a component phase table similar to `argo watch`) that we can contribute to this feature, but we elected to start small to simplify approval.

Relevant issue: https://github.com/kubeflow/pipelines/issues/6792

**Checklist:**
- [x] The title for your pull request (PR) should follow our title convention. [Learn more about the pull request title convention used in this repository](https://github.com/kubeflow/pipelines/blob/master/CONTRIBUTING.md#pull-request-title-convention).

### CODE REVIEW

**Review by connor-mccarthy — COMMENTED** [review]
@droctothorpe, I really like this feature (and personally like `rich` very much). However, I do not think we want to add the associated dependencies.

Is there a standard library progress bar implementation we can pull from?

**Comment by droctothorpe** [comment]
> @droctothorpe, I really like this feature (and personally like `rich` very much). However, I do not think we want to add the associated dependencies.
> 
> Is there a standard library progress bar implementation we can pull from?

Thanks for the lightning quick reply! We can def try to do this with the stdlib. It probably won't be as pretty / colorful, but that's okay. All the rich stuff is encapsulated so it should be pretty easy to swap out.

**Comment by chensun** [comment]
I'm not sure how much value the process percentage would add, yet it could be misleading as the runtime for different components varies and we have no way to predict/estimate that.

**Comment by droctothorpe** [comment]
Agreed. `finished / total` seems more intuitive. Thoughts?

**Comment by droctothorpe** [comment]
@connor-mccarthy We removed the rich dependency and are now constructing the progress bar and loading animation with the stdlib (it was a lot of fun actually, thanks for the suggestion!). It's much cleaner now. @chensun we replaced the percentage indicator with a finished_components / total_components indicator for improved clarity. Let us know if you guys have any other feedback! Thanks!

**Comment by droctothorpe** [comment]
[Here's](https://i.imgur.com/UdhPWRA.gif) a gif that illustrates the visuals.

**Comment by droctothorpe** [comment]
Found an edge case. Will update once it's addressed.

**Comment by droctothorpe** [comment]
Done. Updated `_get_total_components` to extract component / task count from the deserialized component root field.

**Comment by droctothorpe** [comment]
Thanks for the greenlight, @connor-mccarthy! The `kubeflow-pipelines-sdk-execution-tests` failures seem unrelated to this PR. Triggered another run in Prow to verify whether or not it's a one-off.

**Review by connor-mccarthy — COMMENTED** [review]
Hey, @droctothorpe. I'm trying to understand how meaningful the completed/total ratio is here. We know that the ratio is not representative of proportional time to pipeline completion, since each individual component can have arbitrary time to completion. Caching will also contribute to the ratio overstating the total time to completion, on average. Additionally, the dynamic nature of pipelines further erodes the meaning of this ratio:

* How should `dsl.Condition` groups be handled? If a conditional branch has 10 components, how should we calculate the total number of components given that the conditional expression value will not be known until runtime?
* What about `dsl.ParallelFor` groups where each iteration contains a DAG and the number of elements to iterate over is a runtime value?
* How do we treat [pipelines as a task](https://www.kubeflow.org/docs/components/pipelines/v2/author-a-pipeline/component-io/#pipeline-io)? Is that one task or the total number of primitive tasks therein?

Given that the meaning of this ratio isn't particularly intuitive, I'm curious what the value of this feature is over the existing `Client.wait_for_run_completion` method.

**Comment by droctothorpe** [comment]
Hey, @connor-mccarthy!

All great questions. Internally, we went back and forth about how to optimize the UX for this. Our preliminary implementation was actually much closer to what the Argo CLI [watch](https://argoproj.github.io/argo-workflows/cli/argo_watch/) command does. The progress indicator was a bit of an afterthought and we elected to start there wrt upstream contribution since it was a smaller PR (and hopefully an easier review 😂 ).

> How should dsl.Condition groups be handled? If a conditional branch has 10 components, how should we calculate the total number of components given that the conditional expression value will not be known until runtime?

Suppose there's a pipeline with 20 components and 10 are conditional and the condition is never met. In that scenario, the completed ratio would be `10 / 20`, but because the overall phase would enter a non `Pending` or `Running` phase, execution would stop (we anticipated and tested several edge cases like that). 

> - What about dsl.ParallelFor groups where each iteration contains a DAG and the number of elements to iterate over is a runtime value?
> - How do we treat [pipelines as a task](https://www.kubeflow.org/docs/components/pipelines/v2/author-a-pipeline/component-io/#pipeline-io)? Is that one task or the total number of primitive tasks therein?

I think that depends on whether or not [components-root](https://github.com/kubeflow/pipelines/pull/8253/files#diff-eb603c6f0c2f5584a0b97a4ff81bfcbd8ec7ed8c3a5b5eb8a3747deb8d274fbdR1397) stores a record of all components. In our testing, that appeared to be the case, but we didn't test against `parallelFor` and pipelines as a task. Do you happen to know off hand? If not, we can definitely test it ourselves. Those could definitely be edge cases that we haven't accounted for. Thanks for bringing them to our attention! It's almost like this is a complicated domain space 🙃 ! That's why we love it ofc.

One alternative is to lean into what Argo Workflows does and just consume the `progress` field. The total will change over time but it should, in theory, provide a reliable indication since total component count is calculated at runtime / evolves over time. The only drawback is that it includes all of the auxiliary nodes that the emissary executor spawns, so the total count will significantly exceed the actual component count. The auxiliary nodes are concealed by the UI and we wanted to provide a UX that was consistent with the UI, but maybe that's not necessary.

We're flexible / totally open to input / suggestions. Just conveying some of the discussion / rationale that led to the current implementation. Thanks for all the insight!

**Comment by connor-mccarthy** [comment]
Thanks, @droctothorpe. A couple questions to help me understand:

> the completed ratio would be 10 / 20

Would this be the _final_ completed ratio when the pipeline finishes running?

> because the overall phase would enter a non Pending or Running phase, execution would stop (we anticipated and tested several edge cases like that).

Can you describe this a little more? What is the "overall phase"? I'm catching up on the nitty gritty of this.

> we didn't test against parallelFor and pipelines as a task. Do you happen to know off hand?

I do not, sorry! In PipelineSpec IR, `ParallelFor` groups, `Condition` groups, and pipelines as a task are represented as DAG components, which themselves have multiple primitive tasks (e.g., [condition](https://github.com/kubeflow/pipelines/blob/58303b6caa490dd1efaf33bab32c62dfd85f8b16/sdk/python/test_data/pipelines/pipeline_with_condition.yaml#L3). I don't know how this maps to Argo workflow template.

> The auxiliary nodes are concealed by the UI and we wanted to provide a UX that was consistent with the UI

I think this is a good idea. A guiding principle for developing the SDK client has been to strive for feature parity and consistency with the UI. I think consistency in a feature like this is particularly important, to make the UX as easy intuitive as possible by maintaining the same abstraction of a single DAG node in both places.

I wish I had some better ideas for this. This is an exciting feature, but I don't think we want to merge until many of these open questions are resolved. IMO, it would be better to wait to (or possibly not) implement this than to implement something that changes in the future and/or confuses users, particularly since the more granular UI is available and resolves many of these concerns.

**Comment by droctothorpe** [comment]
> Would this be the final completed ratio when the pipeline finishes running?

Yep!

> Can you describe this a little more? What is the "overall phase"? I'm catching up on the nitty gritty of this.

This is in reference to the `phase` field of Argo Workflow's [WorkflowStatus](https://argoproj.github.io/argo-workflows/fields/#workflowstatus) struct. [Here](https://github.com/argoproj/argo-workflows/blob/47544cc02a8663b5b69e4c213a382ff156deb63e/pkg/apis/workflow/v1alpha1/workflow_types.go#L50-L66) are the different phases in AWF's source code.

Our initial implementation stopped execution when `finished` [components] was equal to `total` components, but we refactored to rely on `phase` instead to simplify the logic and account for edge cases.

[This block](https://github.com/kubeflow/pipelines/pull/8253/files#diff-eb603c6f0c2f5584a0b97a4ff81bfcbd8ec7ed8c3a5b5eb8a3747deb8d274fbdR1470-R1486) stops execution of the progress indicator whenever `phase` is neither `Pending` nor `Running`, regardless of the `completed` / `total` component ratio.

We can definitely test how total component count is calculated for PipelineSpec IR and ParallelFor to validate compatibility.

> IMO, it would be better to wait to (or possibly not) implement this than to implement something that changes in the future and/or confuses users, particularly since the more granular UI is available and resolves many of these concerns.

Agreed! We'll dig deeper. It might take a little while since we've pivoted to writing unit tests for our subclass of Client and plan to PR unit tests upstream before looping back around to this.

File: sdk/python/kfp/client/client.py [hunk 1]
```diff
@@ -1338,6 +1341,150 @@ def wait_for_run_completion(self, run_id: str,
...
+        status = workflow_manifest.get("status", {})
+        nodes = status.get("nodes", {})
+
+        completed = 0
+        for _, value in nodes.items():
+            # We only want to include nodes that execute the user-defined component
+            # code. The auxiliary nodes are implementation details that users are likely
+            # oblivious about and would be confused by.
+            # Source: https://github.com/kubeflow/pipelines/blob/master/backend/src/v2/compiler/argocompiler/container.go#L167
+            if value["templateName"] == "system-container-impl":
```
Comment (benjamintanweihao) [inline sdk/python/kfp/client/client.py:1424]: In this case the condition should be flipped right?

```
 if value["templateName"] != "system-container-impl":
```
